### PR TITLE
ENH: `differentiate.hessian`: use `jacobian` to compute accurate Hessian

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Follow-up to gh-20630

#### What does this implement/fix?
Adds a function to compute accurate Hessians using a finite difference approximation. Vectorized to compute Hessian at an arbitrary number of points in a single call.

#### Additional information
To do eventually:
- [x] Test (and fix, where applicable) attributes of return object (for this and `_jacobian`). It doesn't look like `nfev` is correct. *Fixed: only sum the inner `nfev` counts for the first `nit` outer iterations.*
- [x] ~~Improve error estimate by considering error of `jacobian`.~~ Just tighten tolerance of inner Jacobian. In the future we can use a legit 2-D stencil.
- [ ] Consider adding features to help user vectorize with respect to multiple parameters. (Separate PR; not essential.)
- [ ] Consider using first Jacobian evaluation to estimate appropriate step size. This could reduce number of iterations required in subsequent evaluations. (Separate PR; not essential.)

### CODE REVIEW

File: scipy/optimize/_differentiate.py [hunk 1]
```diff
@@ -713,35 +717,40 @@ def _jacobian(func, x, *, atol=None, rtol=None, maxiter=10,
...
-            ``-3`` : A non-finite value was encountered.
-            ``-4`` : Iteration was terminated by `callback`.
-            ``1`` : The algorithm is proceeding normally (in `callback` only).
+
+            - ``0`` : The algorithm converged to the specified tolerances.
+            - ``-1`` : The error estimate increased, so iteration was terminated.
+            - ``-2`` : The maximum number of iterations was reached.
+            - ``-3`` : A non-finite value was encountered.
+            - ``-4`` : Iteration was terminated by `callback`.
+            - ``1`` : The algorithm is proceeding normally (in `callback` only).
```
Comment (mdhaber) [inline scipy/optimize/_differentiate.py:733]: Not using callback right now.
```suggestion
```

File: scipy/optimize/_differentiate.py [hunk 2]
```diff
@@ -854,3 +862,187 @@ def wrapped(x):
...
+    nfev = []  # track inner function evaluations
+    res = _jacobian(df, x, rtol=rtol, **kwargs)  # jacobian of jacobian
+
+    nfev = np.cumsum(nfev, axis=0)
+    res.nfev = np.take_along_axis(nfev, res.nit[np.newaxis, ...], axis=0)[0]
+    res.ddf = res.df
+    del res.df  # this is renamed to ddf
+    del res.nit  # this is only the outer-jacobian nit
+
+    i = res.error <= np.swapaxes(res.error, 0, 1)
```
Comment (mdhaber) [inline scipy/optimize/_differentiate.py:1042]: I think the failure occurs when the errors are equal there needs to be an additional condition to ensure that corresponding off-diagonals are mutually exclusive.

File: scipy/differentiate/_differentiate.py [hunk 3]
```diff
@@ -755,23 +762,26 @@ def jacobian(f, x, *, tolerances=None, maxiter=10,
...
-            - ``-4`` : Iteration was terminated by `callback`.
 
         df : float array
             The Jacobian of `f` at `x`, if the algorithm terminated
             successfully.
         error : float array
             An estimate of the error: the magnitude of the difference between
-            the current estimate of the derivative and the estimate in the
-            previous iteration.
+            the current estimate and the estimate in the previous iteration.
```
Comment (mdhaber) [inline scipy/differentiate/_differentiate.py:771]: Changed inadvertently during merge conflict resolution.
```suggestion
            the current estimate of the derivative and the estimate in the
            previous iteration.
```

File: scipy/differentiate/tests/test_differentiate.py [hunk 4]
```diff
@@ -497,9 +530,9 @@ def test_examples(self, size, func):
         rng = np.random.default_rng(458912319542)
         m, n = func.mn
         x = rng.random(size=(m,) + size)
-        res = jacobian(func, x).df
+        res = jacobian(func, x)
         ref = func.ref(x)
-        np.testing.assert_allclose(res, ref, atol=1e-10)
+        np.testing.assert_allclose(res.df, ref, atol=1e-10)
 
     def test_iv(self):
```
Comment (mdhaber) [inline scipy/differentiate/tests/test_differentiate.py:537]: Looks like this can be removed.

**Review by mdhaber — COMMENTED** [review]
Need to take another pass through documentation - after making `differentiate` and `jacobian` public, there may be other changes that should be reflected here. *Update: done. I also removed the attempt to enforce symmetry since we couldn't nail down how function evaluations should be counted in that case - of course, I could also just remove `nit` from the results, in which case `ddf` and `error` are easy to make symmetric. If we're happy with the approach for the error estimate, I think this is ready for review.*

**Comment by stefanv** [comment]
This is neat! :sunglasses:  I read through the code and noticed no obvious problems, but I am also not an expert on numerical Hessian calculation.

**Comment by j-bowhay** [comment]
If you resolve the merge conflict I would be happy to review this next week

File: scipy/differentiate/_differentiate.py [hunk 5]
```diff
@@ -889,3 +900,193 @@ def wrapped(x):
...
+
+    See Also
+    --------
+    differentiate, jacobian
+
+    Notes
+    -----
+    Suppose we wish to evaluate the Hessian of a function
+    :math:`f: \mathbf{R^m} \rightarrow \mathbf{R}`, and we assign to variable
+    ``m`` the positive integer value of :math:`m`. If we wish to evaluate
```
Comment (j-bowhay) [inline scipy/differentiate/_differentiate.py:992]: nit but I would find this easier to read as it means $m$ is consistently typeset
```suggestion
    :math:`f: \mathbf{R}^m \rightarrow \mathbf{R}`, and we assign to variable
    ``m`` the positive integer value of :math:`m`. If we wish to evaluate
```

Comment (j-bowhay) [reply scipy/differentiate/_differentiate.py:992]: I think ideally the same change for `jacobian` too

File: scipy/differentiate/_differentiate.py [hunk 6]
```diff
@@ -889,3 +900,193 @@ def wrapped(x):
...
+    emitted otherwise.
+
+    References
+    ----------
+    .. [1] Hessian matrix, *Wikipedia*,
+           https://en.wikipedia.org/wiki/Hessian_matrix
+
+    Examples
+    --------
+    The Rosenbrock function maps from :math:`\mathbf{R}^m \righarrow \mathbf{R}`;
```
Comment (j-bowhay) [inline scipy/differentiate/_differentiate.py:1027]:
```suggestion
    The Rosenbrock function maps from :math:`\mathbf{R}^m \rightarrow \mathbf{R}`;
```

File: scipy/differentiate/_differentiate.py [hunk 7]
```diff
@@ -889,3 +900,193 @@ def wrapped(x):
...
+    message = (f"The specified `{rtol=}`, but error estimates are likely to be "
+               f"unreliable when `rtol < {rtol_min}`.")
+    if 0 < rtol < rtol_min:  # rtol <= 0 is an error
+        warnings.warn(message, RuntimeWarning, stacklevel=2)
+        rtol = rtol_min
+
+    def df(x):
+        tolerances = dict(rtol=rtol/100, atol=atol)
+        temp = jacobian(f, x, tolerances=tolerances, **kwargs)
+        nfev.append(temp.nfev if len(nfev)==0 else temp.nfev.sum(axis=-1))
```
Comment (j-bowhay) [inline scipy/differentiate/_differentiate.py:1080]:
```suggestion
        nfev.append(temp.nfev if len(nfev) == 0 else temp.nfev.sum(axis=-1))
```

File: scipy/differentiate/tests/test_differentiate.py [hunk 8]
```diff
@@ -446,7 +446,38 @@ def test_saddle_gh18811(self, case):
...
+class JacobianHessianTest:
+    def test_iv(self):
+        jh_func = self.jh_func.__func__
+
+        # Test input validation
+        message = "Argument `x` must be at least 1-D."
+        with pytest.raises(ValueError, match=message):
+            jh_func(np.sin, 1, tolerances=dict(atol=-1))
+
+        # Confirm that other parameters are being passed to `_derivative`,
```
Comment (j-bowhay) [inline scipy/differentiate/tests/test_differentiate.py:458]: Not sure what  `` `_derivative` `` refers to?

**Review by j-bowhay — COMMENTED** [review]
Some minor nits but otherwise in good shape

File: scipy/differentiate/tests/test_differentiate.py (hunk 8 above)
Comment (mdhaber) [reply scipy/differentiate/tests/test_differentiate.py:458]: It used to be private. 
```suggestion
        # Confirm that other parameters are being passed to `derivative`,
```

File: scipy/differentiate/_differentiate.py (hunk 5 above)
Comment (mdhaber) [reply scipy/differentiate/_differentiate.py:992]: Yup, I think I got them all. (locally; will push soon.)

File: scipy/differentiate/tests/test_differentiate.py (hunk 8 above)
Comment (j-bowhay) [reply scipy/differentiate/tests/test_differentiate.py]: `derivative` is also nice in that it is clearer (to me at least) it is elementwise and that it is not the gradient

Comment (mdhaber) [reply scipy/differentiate/tests/test_differentiate.py]: OK, I'll open a separate PR after this one.

Comment (j-bowhay) [reply scipy/differentiate/tests/test_differentiate.py]: Julia also favors `derivative`
https://docs.sciml.ai/FiniteDiff/stable/#Scalar-Derivatives
https://juliadiff.org/ForwardDiff.jl/stable/user/api/#Derivatives-of-f(x::Real)::Union{Real,AbstractArray}

**Comment by mdhaber** [comment]
OK @j-bowhay I think the doc issues you mentioned were resolved. I will open a separate PR to make it `differentiate.derivative` because they will need an email to the list.

**Review by j-bowhay — APPROVED** [review]
LGTM thanks!

**Pull request merged**